incubator/hnc/pkg/controllers/hierarchy_controller.go

```diff
@@ -296,64 +301,82 @@ func (r *HierarchyReconciler) syncLabel(log logr.Logger, nsInst *corev1.Namespac
 // have been marked as required. If any required children are missing, we add them to the in-memory
 // forest and enqueue the (missing) child for reconciliation; we also handle various error cases.
 func (r *HierarchyReconciler) syncChildren(log logr.Logger, inst *api.HierarchyConfiguration, ns *forest.Namespace) {
+	// Update the most recent list of children from the forest
 	inst.Status.Children = ns.ChildNames()
+
+	// We'll reset any of these conditions if they occur.
+	ns.ClearAllConditions(api.RequiredChildConflict)
+
 	// Make a set to make it easy to look up if a child is required or not
-	reqSet := map[string]bool{}
+	isRequired := map[string]bool{}
 	for _, r := range inst.Spec.RequiredChildren {
-		reqSet[r] = true
+		isRequired[r] = true
 	}
 
 	// Check the list of actual children against the required children
 	for _, cn := range inst.Status.Children {
 		cns := r.Forest.Get(cn)
-		if _, isRequired := reqSet[cn]; isRequired {
-			// This is a required child of this namespace
-			cns.RequiredChildOf = ns.Name()         // mark in in the forest
-			delete(reqSet, cn)                      // remove so we know we found it
-			if cns.Exists() && cns.Parent() != ns { // condition if it's assigned elsewhere
-				msg := fmt.Sprintf("required subnamespace %s exists but has parent %s", cn, cns.Parent().Name())
-				ns.SetCondition(forest.Local, api.CritRequiredChildConflict, msg)
-			}
-		} else if cns.RequiredChildOf == ns.Name() {
+
+		if cns.RequiredChildOf != "" && cns.RequiredChildOf != ns.Name() {
+			// Since the list of children of this namespace came from the forest, this implies that the
+			// in-memory forest is out of sync with itself: requiredChildOf != parent. Obviously, this
+			// should never happen.
+			//
+			// Let's just log an error and enqueue the other namespace so it can report the condition.
+			// Also, reset the forest to match reality.
+			log.Error(forest.OutOfSync, "While syncing children", "child", cn, "requiredChildOf", cns.RequiredChildOf)
+			r.enqueueAffected(log, "forest out-of-sync: requiredChildOf != parent", cns.RequiredChildOf)
+			cns.RequiredChildOf = ""
```

Comment:
We may not need this line? Since it will be set in the following `if/else`?

Reply:
Good point. This code used to be in a different location.